packages/@aws-cdk/aws-eks/lib/k8s-resource.ts

```diff
@@ -51,23 +56,69 @@ export class KubernetesResource extends Construct {
   constructor(scope: Construct, id: string, props: KubernetesResourceProps) {
     super(scope, id);
 
-    const stack = Stack.of(this);
-
-    // we maintain a single manifest custom resource handler for each cluster
-    const handler = props.cluster._k8sResourceHandler;
-    if (!handler) {
+    if (!props.cluster._clusterResource) {
       throw new Error(`Cannot define a KubernetesManifest resource on a cluster with kubectl disabled`);
     }
 
-    new cfn.CustomResource(this, 'Resource', {
-      provider: cfn.CustomResourceProvider.lambda(handler),
+    const stack = Stack.of(this);
+    const provider = KubernetesResourceProvider.getOrCreate(this);
+
+    new CustomResource(this, 'Resource', {
+      provider: provider.provider,
       resourceType: KubernetesResource.RESOURCE_TYPE,
       properties: {
         // `toJsonString` enables embedding CDK tokens in the manifest and will
         // render a CloudFormation-compatible JSON string (similar to
         // StepFunctions, CloudWatch Dashboards etc).
         Manifest: stack.toJsonString(props.manifest),
+        ClusterName: props.cluster.clusterName,
+        RoleArn: props.cluster._clusterResource.getCreationRoleArn(provider.role)
       }
     });
   }
 }
+
+class KubernetesResourceProvider extends NestedStack {
+  /**
+   * Creates a stack-singleton resource provider nested stack.
+   */
+  public static getOrCreate(scope: Construct) {
+    const stack = Stack.of(scope);
+    const uid = '@aws-cdk/aws-eks.KubernetesResourceProvider';
+    return stack.node.tryFindChild(uid) as KubernetesResourceProvider || new KubernetesResourceProvider(stack, uid);
+  }
+
+  /**
+   * The custom resource provider.
+   */
+  public readonly provider: cr.Provider;
+
+  /**
+   * The IAM role used to execute this provider.
+   */
+  public readonly role: iam.IRole;
+
+  constructor(scope: Construct, id: string) {
```

Comment:
As before, this should probably be `private`.

Reply:
Done